relay-system/src/service.rs

```diff
@@ -352,15 +365,78 @@ impl<I: Interface> Addr<I> {
 ///
 /// Instances are created automatically when [spawning](Service::spawn_handler) a service, or can be
 /// created through [`channel`]. The channel closes when all associated [`Addr`]s are dropped.
-pub type Receiver<I> = mpsc::UnboundedReceiver<I>;
+pub struct Receiver<I: Interface> {
+    rx: mpsc::UnboundedReceiver<I>,
+    name: &'static str,
+    interval: tokio::time::Interval,
+    backlog: Arc<AtomicU64>,
+}
+
+impl<I: Interface> Receiver<I> {
+    /// Receives the next value for this receiver.
+    ///
+    /// This method returns `None` if the channel has been closed and there are
+    /// no remaining messages in the channel's buffer. This indicates that no
+    /// further values can ever be received from this `Receiver`. The channel is
+    /// closed when all senders have been dropped.
+    ///
+    /// If there are no messages in the channel's buffer, but the channel has
+    /// not yet been closed, this method will sleep until a message is sent or
+    /// the channel is closed.
+    pub async fn recv(&mut self) -> Option<I> {
+        loop {
+            tokio::select! {
+                biased;
+
+                _ = self.interval.tick() => {
```

Comment:
I'm not sure about this select now. We will be missing some metrics, since once we get the message we exit this receive method. 

I'm thinking maybe it will be better to spawn the `tokio::task` and just report the metrics from there.

Reply:
`recv` is called in a loop by the service, so this will run continuously. All that this inner loop and `select!` does is inject into the existing external loop and slide some metrics in.